[pytorch] continue to rewrite gen_python_functions.py with typed models

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47438 [pytorch] split out trace type generator and migrate to new codegen model
* #47437 [pytorch] factor out return_names api
* #47436 [pytorch] TraceType codegen tweak - newline before redispatch call
* **#46978 [pytorch] continue to rewrite gen_python_functions.py with typed models**

Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

Differential Revision: [D24589210](https://our.internmc.facebook.com/intern/diff/D24589210)

ljk53 committed:
[pytorch] continue to rewrite gen_python_functions.py with typed models

Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] continue to rewrite gen_python_functions.py with typed models"


Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

[ghstack-poisoned]

ljk53 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -483,130 +452,224 @@ def is_noarg_binding(overloads):
+    """
+    if overload.outplace is not None:
+        # dispatch output and no-output variants, branch on _r.isNone(<out_idx>)
+        return PY_VARIABLE_OUT.substitute(
```
I noticed that the new codegen is moving away from CodeTemplate - but the auto-indent feature is useful in this case as the emit_single_dispatch can have different indents: inside switch-case, inside if-else, inside both or none.

ezyang replied:
Yes, it is troublesome. For now, I have have been opting for code gen readability (f-strings) at the expense of generated code readability (yolo indentation) but generated code readability is getting worse and worse.  I'm kind of pissed that Python doesn't have some sort of pluggable f-string mechanism for this case.

I can only really think of a few options for this problem:
1. Run some sort of formatter (clang-format, maybe?) on the generated code after we generate it. Sounds slow.
2. Manually annotate multiline string formats with an indent level that triggers indentation, e.g., 

```
void f() {{
    {multiline(thing, 4)}
}}
```

3. Stop using f-strings

ljk53 committed:
Update on "[pytorch] continue to rewrite gen_python_functions.py with typed models"


Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

Differential Revision: [D24589210](https://our.internmc.facebook.com/intern/diff/D24589210)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] continue to rewrite gen_python_functions.py with typed models"


Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

Differential Revision: [D24589210](https://our.internmc.facebook.com/intern/diff/D24589210)

[ghstack-poisoned]

ezyang commented:
What's the general approach I should take for reviewing this PR?

ljk53 committed:
Update on "[pytorch] continue to rewrite gen_python_functions.py with typed models"


Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

Differential Revision: [D24589210](https://our.internmc.facebook.com/intern/diff/D24589210)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] continue to rewrite gen_python_functions.py with typed models"


Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

Differential Revision: [D24589210](https://our.internmc.facebook.com/intern/diff/D24589210)

[ghstack-poisoned]

ljk53 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
+    for sig, out in outplaces.items():
+        if sig not in bases:
+            candidates: List[str] = []
+            for overload in overloads:
```
merged with #47087 and tested by hacking the native_functions.yaml:
```
RuntimeError: While identifying overloads, we found an out schema addmv(Tensor input, Tensor mat, Tensor vec, *, Scalar beta=1, Scalar alpha=1, Tensor out=None) without a corresponding non-out variant. We expected the non-out variant to have schema: 
- addmv(Tensor input, Tensor mat, Tensor vec, *, Scalar beta=1, Scalar alpha=1)
Please check that you spelled the schema correctly in native_functions.yaml. We discovered the following candidate(s): 
- addmv(Tensor input, Tensor mat, Tensor vec, *, Scalar beta=1)
```

ezyang commented:
Last week, @ljk53 and @bhosmer and I had a brief discussion about how to go about reviewing this PR. Big take away: Phabricator does a better job of generating diff than GitHub, so go look at the diff there (it identifies code motion).

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
+        # TODO: should use some canonical form instead of 'str(arg.type)' - see comments
+        # above. The old codegen used the deprecated 'dynamic_type(arg.type)', which
+        # ignores the optional annotation, i.e. 'Scalar' and 'Scalar?'.
+        any_unequal = any(str(arg1.type) != str(arg2.type)
```
You know you can just test these for equality right? So no need to stringify them first. (Stringifying is poor man's pattern matching.)

ezyang replied:
In fact, you can't you literally just do `args1 != args2` because you've already checked that the lengths are the same above?

ezyang replied:
Oh you can't, because you only want to check the types

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
+                                   or is_arg_smaller(arg1.type, arg2.type)
                                    for arg1, arg2 in zip(args1, args2))
-        return any_smaller and all_smaller_or_equal
+        return any_unequal and all_smaller_or_equal
```
I understand this is the fault of the original code, but I just want to remark: the use of "any" and "all" here obscures the essential logical property going on here: if you know how to compute `lte` and `eq`, you can compute `lt = lte && !eq`. The swapping between any/all is the result of pushing the negation inside of the all quantifier, but the quantifier doesn't really have anything to do with the logical computation going on here.

ezyang replied:
Also, (once again, artifact of the old code) I'm a little perplexed as to why equality matters here, because I don't feel like there really should be any overloads in the list with exactly the same types, so given two distinct signatures they should always be unequal.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
-                format(declaration['name']))
-        return name
+                del larger_than[j]
+                sorted_ids.append(j)
```
Also the old code's fault, but mutating lists as I iterate over them always gives me the jeebies

ljk53 replied:
> Also the old code's fault, but mutating lists as I iterate over them always gives me the jeebies

ok - I've simply changed it to iterating through index indirectly instead of sorted_ids directly - not sure it hurts readability :-)

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
-    """
-    grouped = defaultdict(dict)
+def group_overloads(
+    overloads: Tuple[PythonSignatureNativeFunctionPair, ...]
```
Nit: this is probably OK just being a `Sequence`

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
+    for overload in overloads:
+        sig = overload.signature.signature_str(skip_outputs=True)
+        if overload.function.func.is_out_fn():
+            outplaces[sig] = overload
```
`assert sig not in outplaces`? Similar with `bases` below.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
+    for sig, base in bases.items():
+        outplace = outplaces.get(sig)
+        grouped.append(PythonSignatureGroup(
+            # prefer the signature with optional out=... arguments
```
Nothing to do here, but I hate this. Will be kind of hard to fix though.

The opposite invariant (use the signature with functional arguments) would be easier to get to; I doubt there are many out= functions that don't have a functional version. Something that isn't so clear just looking at the comment here or in PythonSignatureGroup comments is *why* we need to prefer the out= variant.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
-        def_template = PY_VARIABLE_METHOD_BINOP_DEF
+        # PyMethodDef entry for binary op, throws not implemented error
+        return f"""\
+{{"{name}", {pyfunc_cast}(TypeError_to_NotImplemented_<{pycname}>), {flags}, NULL}},"""
```
Thanks for inlining these, I appreciate it.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -245,190 +239,157 @@ def group_declarations_by_op_name(declarations):
-        py_forwards.extend(forward_decls(name, overload_decls, is_python_method, module))
+        overloads = tuple(decl_to_signature_function_pair(decl, method=is_python_method)
+                          for decl in python_functions[name])
+        noarg = len(overloads) == 1 and overloads[0].signature.arguments_count() == 0
```
Note to self: wondering what a noarg is and why we have to compute here, rather than inside the functions.

ljk53 replied:
no particular reason - just that this way we can avoid passing 'overloads' to 'method_def' / 'forward_decls' because this bit is what all they need.
I've changed it back to passing overloads instead to make it more uniform.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -245,190 +239,157 @@ def group_declarations_by_op_name(declarations):
+
+@with_native_function
+def gen_namedtuple_typename_key(f: NativeFunction) -> str:
+    name = cpp.name(f.func)
```
cpp.name? Really?

ezyang replied:
Hmm, this is probably an internal name, judging from the algo below

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -245,190 +239,157 @@ def group_declarations_by_op_name(declarations):
+
+def emit_namedtuple_typedefs(
+    overloads: Tuple[PythonSignatureNativeFunctionPair, ...]
+) -> Tuple[List[str], Dict[str, str]]:
```
Getting pretty bad str blindness here

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -245,190 +239,157 @@ def group_declarations_by_op_name(declarations):
-    for decl in declarations:
-        fieldnames = namedtuple_fieldnames(decl)
-        if fieldnames == []:
-            decl['namedtuple_typeref'] = ''
```
GOOD RIDDANCE, ha!

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -245,190 +239,157 @@ def group_declarations_by_op_name(declarations):
-        typename = typenames.get(key)
+        name = cpp.name(overload.function.func)  # use @with_native_function?
+        tn_key = gen_namedtuple_typename_key(overload.function)
+        typename = typenames.get(tn_key)
```
It would make me feel better if there was some asserts here that would tell me that order of overloads doesn't matter for these initializations but I can't think of an easy way to actually do this.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -483,130 +453,166 @@ def is_noarg_binding(overloads):
-    method_footer = ['END_HANDLE_TH_ERRORS']
+    # NOTE: we type the unpacked self as Tensor not Variable to avoid return type
+    # discrepancies on method resolution (e.g. Variable::detach_ returns void
+    # rather than Tensor &)
```
BTW, this comment is defunct now, because Tensor and Variable are the same in C++ now

ljk53 replied:
ok, I deleted this comment - not worth calling out anymore?

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -483,130 +453,166 @@ def is_noarg_binding(overloads):
+if(check_has_torch_function(self_)) {{
+  return handle_torch_function(self_, "{name}");
+}}
+""" if method else ''
```
nit: the `if method` has the same indentation as `if noarg` so I nearly thought it was acutally on the next line haha. Might be easier to read if we just have another nested if block inside

ljk53 replied:
done

ljk53 committed:
Update on "[pytorch] continue to rewrite gen_python_functions.py with typed models"


Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

Differential Revision: [D24589210](https://our.internmc.facebook.com/intern/diff/D24589210)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] continue to rewrite gen_python_functions.py with typed models"


Refactored and added type annotations to the most part of the file.

Some top-level codegen functions are called by other codegen scripts.
Will migrate them in subsequent PRs.

Differential Revision: [D24589210](https://our.internmc.facebook.com/intern/diff/D24589210)

[ghstack-poisoned]
